content/posts/python-type.md

```diff
@@ -0,0 +1,143 @@
++++
+title = "Objects in Python"
+date = 2023-04-03
+tags = ["python"]
++++
+
+Everything in Python has a type (even the type `type`!) because everything is an object.
+
+Objects are abstraction of data in Python.
+All data in a Python program is stored as a object.
+
+An object allows instances of it to be created.
+The created instances, which are also objects are destroyed by garbage collection.
+
+An object has three properties - identity, value, type.
+The `id` of an object never changes once an object is created.
+`id(x)` returns an unique identity to the object `x`. 
+
+When the value of object is changeable, the object is said to be mutable and when the value is not changeable, the object is said to be immutable.
+
+## `type` object
+
+`type` is an object which can be used to create other objects.
+For example, the `type` object is used internally to create `class` objects.
+An user-defined class is an object of type `type`.
+It is a new kind of type (instance) of the type class.
+
+The type of `type` object is `type`.
+For `type` object, the type is an identity.
+Example: `print (type(type))` prints `<class 'type'>`
+
+## `function` object
+
+A function is also another instance of `type` object.
+When we define a function, the defined function is an instance of the function object.
+
+Example:
+```
+def foo():
+    return
+
+print (type(foo))  # <class 'function'>
+print (type(type(foo)))  # <class 'type'>
+```
+
+## `class` object
+In python, all classes are themselves instances of other classes.
+`class` objects are instances of `type` class.
+A class also allows to makes instances of it's type.
+
+
+```python3
+class Foo:
+    pass
+
+print (type(Foo)) # <class 'type'>
+```
+
+### Instance method object
+
+A instance method object is same as a function object but it accepts a class instance as a first argument.
+It combines a class, a class instance and any callable user-defined function.
+
+Example:
+
+```python3
+class Foo():
+    def print_self(self):
+         print ('type ': self, 'id ': id(self))
+
+f = Foo()
+print (type(f))  # output: type <class '__main__.Foo'> id 4408722096
+f.print_self()  # output: type <class '__main__.Foo'> id 4408722096
```

Comment:
Maybe make a note that the id is the same because it is the same instance, but if it is a different instance, say g = Foo(), it will be a different id